src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/visitorders/VisitBalanceMigrationService.kt

```diff
@@ -0,0 +1,119 @@
+package uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visitorders
+
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.data.domain.PageImpl
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.data.MigrationContext
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.helpers.trackEvent
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.listeners.MigrationMessageType
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.nomismappings.model.VisitBalanceMappingDto
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.nomismappings.model.VisitBalanceMappingDto.MappingType.MIGRATED
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.nomisprisoner.model.PrisonerId
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.nomisprisoner.model.PrisonerVisitOrderBalanceResponse
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.MigrationService
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.MigrationType
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.NomisApiService
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visit.balance.model.VisitAllocationPrisonerMigrationDto
+import java.time.LocalDate
+
+@Service
+class VisitBalanceMigrationService(
+  val nomisApiService: NomisApiService,
+  val visitBalanceMappingService: VisitBalanceMappingApiService,
+  val visitBalanceNomisApiService: VisitBalanceNomisApiService,
+  val dpsApiService: VisitBalanceDpsApiService,
+  @Value("\${visitbalance.page.size:1000}") pageSize: Long,
+  @Value("\${complete-check.delay-seconds}") completeCheckDelaySeconds: Int,
+  @Value("\${complete-check.count}") completeCheckCount: Int,
+) : MigrationService<VisitBalanceMigrationFilter, PrisonerId, VisitBalanceMappingDto>(
+  mappingService = visitBalanceMappingService,
+  migrationType = MigrationType.VISIT_BALANCE,
+  pageSize = pageSize,
+  completeCheckDelaySeconds = completeCheckDelaySeconds,
+  completeCheckCount = completeCheckCount,
+) {
+  private companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  // TODO Set up filtering
+  override suspend fun getIds(
+    migrationFilter: VisitBalanceMigrationFilter,
+    pageSize: Long,
+    pageNumber: Long,
+  ): PageImpl<PrisonerId> = nomisApiService.getPrisonerIds(
+    pageNumber = pageNumber,
+    pageSize = pageSize,
+  )
+
+  override suspend fun migrateNomisEntity(context: MigrationContext<PrisonerId>) {
+    val nomisPrisonNumber = context.body.offenderNo
+    val alreadyMigratedMapping = visitBalanceMappingService.getByNomisPrisonNumberOrNull(nomisPrisonNumber)
```

Comment:
This endpoint may change - to ensure we are only migrating those with balance adjustments